cmd/zgenhostid/zgenhostid.c

```diff
@@ -38,44 +38,71 @@
 
 static void usage(void);
 
-char path[PATH_MAX] = "/etc/hostid";
+static void
+usage(void)
+{
+	(void) fprintf(stderr,
+	    "usage: zgenhostid [-fh] [-o path] [value]\n\n"
+	    "  -f\t\t force hostid file write\n"
+	    "  -h\t\t print this usage and exit\n"
+	    "  -o <filename>\t write hostid to this file\n\n"
+	    "If hostid file is not present or invalid, store a hostid in it.\n"
+	    "The optional value must be an 8-digit hex number between"
+	    "1 and 2^32-1.\n"
+	    "If no value is provided, a random one will"
+	    "be generated.\n"
+	    "The value must be unique among your systems.\n");
+	exit(EXIT_FAILURE);
+	/* NOTREACHED */
+}
 
 int
 main(int argc, char **argv)
 {
-	int opt;			// getopt iterator
-	int forcew = 0;			// force  file write
-	uint32_t hostid = 0;		// final hostid value
-	unsigned long int input_i = 0;	// int to hold converted input
-
+	/* default file path, can be optionally set by user */
+	char path[PATH_MAX] = "/etc/hostid";
+	/* holds converted user input or lrand48() generated value */
+	unsigned long input_i = 0;
+
+	int opt;
+	int pathlen;
+	int force_fwrite = 0;
 	while ((opt = getopt_long(argc, argv, "fo:h?", 0, 0)) != -1) {
 		switch (opt) {
 		case 'f':
-			forcew = 1;
+			force_fwrite = 1;
 			break;
 		case 'o':
-			snprintf(path, sizeof (path), "%s", optarg);
+			pathlen = snprintf(path, sizeof (path), "%s", optarg);
+			if (pathlen >= sizeof (path)) {
+				fprintf(stderr, "%s\n", strerror(EOVERFLOW));
+				exit(EXIT_FAILURE);
+			} else if (pathlen < 1) {
+				fprintf(stderr, "%s\n", strerror(EINVAL));
+				exit(EXIT_FAILURE);
+			}
 			break;
 		case 'h':
 		case '?':
 			usage();
 		}
 	}
 
-	if (argv[optind]) {
-		// we expect 0x prefixed or unprefixed input
-		// filter out 6 digit 0x prefixed values
-		char in_s[12];
-		int len = snprintf(in_s, sizeof (in_s), "%s", argv[optind]);
+	char *in_s = argv[optind];
+	if (in_s != NULL) {
+		/*
+		 * we expect 0x prefixed or unprefixed string input
+		 * filter out 6 digit 0x prefixed values
+		 */
+		int len = strlen(in_s);
 		if (len == 8 && strncasecmp("0x", in_s, 2) == 0) {
 			fprintf(stderr, "%s\n", strerror(ERANGE));
 			usage();
 		}
 
-		// we need exactly 8 or 10 (if input is 0x prefixed) characters
+		/* need exactly 8 or 10 (if input is 0x prefixed) characters */
 		const char *hex = "0123456789abcdefxABCDEFX";
-		int numhex = 0;
-		numhex = strspn(in_s, hex);
+		int numhex = strspn(in_s, hex);
 		if (numhex != 8 && numhex != 10) {
```

Comment:
I believe that if the input string starts with 10 hex digits, but has no 0x prefix, strtoul will parse 8 digits of it, potentially leading to confusion.  For example, with string "1111222233".  Since you require they give you exactly 8 digits of hex anyway, maybe just keep track of the number of prefix digits and subtract that from the length, e.g. ```if ((numhex-numprefix) != 8)```.

Reply:
numhex may be == 10, but int range check will not succeed at least on systems where ULONG_MAX is 8 bytes.

but it can misfire on 32bit systems, because `strtoul` will return `ULONG_MAX` which will likely be equal to `UINT32_MAX`
but the *endptr check was supposed to prevent that, but looks like it's not working as it should, because condition defined wrong.

on ppc32
```
./a.out -o /tmp/hostid -f ffffffffff
0000000 ffff ffff
0000004
```
changing `if (errno != 0)` to be only condition properly catches truncation, so logic `if (errno != 0 && *endptr != '\0')` is not working.

I think it should be sufficient for catching overflows, because `errno` will be non-zero if truncated/overflown, so I'll remove endptr check.

thanks for catching!

I can skip first 2 bytes if string starts with `0x` and check exactly for 8 characters, this will simplify things.